File: src/sources/github/policy.ts
import type { ChangedFile } from './gh.js';

export interface PolicyVerdict {
  ok: boolean;
  reasons: string[];
}

/** Blast-radius policy for auto-merge, per github source instance. */
export interface MergePolicy {
  allowedGlobs: string[];
  maxChangedLines: number;
}

/** Minimal glob matcher supporting `**` and `*` — enough for path prefixes. */
function globToRegExp(glob: string): RegExp {
  const escaped = glob
    .replace(/[.+^${}()|[\]\\]/g, '\\$&')
    .replace(/\*\*/g, ' ') // placeholder for **
    .replace(/\*/g, '[^/]*')
    .replace(/ /g, '.*');
  return new RegExp(`^${escaped}$`);
}

/**
 * Decide whether a diff is safe to auto-merge:
 *  - every changed file must match an allowed glob (blast radius), and
 *  - total changed lines must be within the configured limit.
 * A failing verdict means "leave the PR for human review", not "reject".
 */
export function evaluateDiff(files: ChangedFile[], policy: MergePolicy): PolicyVerdict {
  const reasons: string[] = [];
  const allowMatchers = policy.allowedGlobs.map(globToRegExp);
  const isAllowedPath = (file: string): boolean => allowMatchers.some((re) => re.test(file));

  if (files.length === 0) {
    return { ok: false, reasons: ['no changed files'] };
  }

  const outside = files.map((f) => f.file).filter((f) => !isAllowedPath(f));
  if (outside.length > 0) {
    reasons.push(`files outside allowed globs: ${outside.join(', ')}`);
  }

  const totalLines = files.reduce((sum, f) => sum + f.added + f.deleted, 0);
  if (totalLines > policy.maxChangedLines) {
    reasons.push(`diff too large: ${totalLines} lines > limit ${policy.maxChangedLines}`);
  }

  return { ok: reasons.length === 0, reasons };
}

File: agent_loop/github/policy.py
"""Blast-radius policy for auto-merge. Ported from policy.ts.

A failing verdict means "leave the PR for human review", not "reject".
"""

from __future__ import annotations

import re
from dataclasses import dataclass, field

from ..config import MergePolicy
from .gh import ChangedFile


@dataclass
class PolicyVerdict:
    ok: bool
    reasons: list[str] = field(default_factory=list)


def _glob_to_regexp(glob: str) -> re.Pattern[str]:
    """Minimal glob → regex supporting ** and * — enough for path prefixes."""
    escaped = re.escape(glob)
    # re.escape turns * into \*, ** into \*\*. Restore glob semantics:
    escaped = escaped.replace(r"\*\*", "\x00")  # placeholder for **
    escaped = escaped.replace(r"\*", "[^/]*")
    escaped = escaped.replace("\x00", ".*")
    return re.compile(f"^{escaped}$")


def evaluate_diff(files: list[ChangedFile], policy: MergePolicy) -> PolicyVerdict:
    """Decide whether a diff is safe to auto-merge: every changed file must match
    an allowed glob, and total changed lines must be within the limit."""
    reasons: list[str] = []
    matchers = [_glob_to_regexp(g) for g in policy.allowed_globs]

    def allowed(f: str) -> bool:
        return any(m.match(f) for m in matchers)

    if not files:
        return PolicyVerdict(ok=False, reasons=["no changed files"])

    outside = [f.file for f in files if not allowed(f.file)]
    if outside:
        reasons.append(f"files outside allowed globs: {', '.join(outside)}")

    total = sum(f.added + f.deleted for f in files)
    if total > policy.max_changed_lines:
        reasons.append(f"diff too large: {total} lines > limit {policy.max_changed_lines}")

    return PolicyVerdict(ok=not reasons, reasons=reasons)
